packages/cli/src/commands/deploy/action.ts

```diff
@@ -0,0 +1,63 @@
+import { TxParams } from '@openzeppelin/upgrades';
+import { compile } from '../../models/compiler/Compiler';
+import { fromContractFullName } from '../../utils/naming';
+import ConfigManager from '../../models/config/ConfigManager';
+import NetworkController from '../../models/network/NetworkController';
+import stdout from '../../utils/stdout';
+import { MethodArgType } from '../../prompts/prompt';
+import { parseMultipleArgs } from '../../utils/input';
+import NetworkFile from '../../models/files/NetworkFile';
+import { createAction } from '../create';
+
+import { AbortAction } from '../../register-command';
+
+interface Options {
+  network: string;
+  skipCompile?: boolean;
+  upgradeable?: boolean;
+
+  // The following are not available as CLI flags, and they are only used in tests.
+  networkFile?: NetworkFile;
+  txParams?: TxParams;
+}
+
+export async function preAction(
+  contractName: string | undefined,
+  deployArgs: string[],
+  options: Options,
+): Promise<void> {
+  if (!options.skipCompile) {
+    await compile();
+  }
+
+  if (options.upgradeable === true) {
+    throw new AbortAction(async () => {
+      // Translate arguments to syntax expected by create.
+      options['args'] = deployArgs.join(',');
+      await createAction(contractName, options);
+    });
+  }
+}
+
+export async function action(contractName: string, deployArgs: string[], options: Options): Promise<void> {
+  // this is an interactive prompt, so should be handled by the prompts
+  // if (!(await hasToMigrateProject(network))) process.exit(0);
+
+  const { network, txParams } = options.txParams ? options : await ConfigManager.initNetworkConfiguration(options);
+
+  const { package: packageName, contract: contractAlias } = fromContractFullName(contractName);
+
+  const controller = new NetworkController(network, txParams, options.networkFile);
+
+  const contract = controller.contractManager.getContractClass(packageName, contractAlias);
+  const constructorInputs: MethodArgType[] = contract.schema.abi.find(f => f.type === 'constructor')?.inputs ?? [];
```

Comment:
I agree that it's weird for `action` to have this concern, but I was reluctant to add anything to "`lib`", a.k.a. `upgrades`, and I don't think it's a good idea to use `upgrades` as the "utils" module for CLI. What do you think?

It's true that currently we are using `upgrades` in this way, so arguably I should just stick to that for now, but I would've liked to go in the direction of improving that situation.

Reply:
I'd add this to `upgrades`, so when we split out the `utils` modules, this gets moved there automatically. Otherwise, when we make the split, this code will still be left there.

And yes, we _will_ be doing that code split; probably right after transpiler gets integrated (or even right before!).